File: frontend/src/api/researchLineage.ts
import { readJson } from './client'
import type { LineageDirection, LineageNodeType, ResearchLineageGraph, ResearchLineageSummary } from '../types/researchLineage'

export interface ResearchLineageQuery {
  root_type?: LineageNodeType | null
  root_id?: string | null
  direction?: LineageDirection
  max_depth?: number
  node_types?: LineageNodeType[]
}

export async function getResearchLineage(query: ResearchLineageQuery = {}): Promise<ResearchLineageGraph> {
  const parameters = new URLSearchParams()
  if (query.root_type && query.root_id) {
    parameters.set('root_type', query.root_type)
    parameters.set('root_id', query.root_id)
  }
  if (query.direction) parameters.set('direction', query.direction)
  if (query.max_depth) parameters.set('max_depth', String(query.max_depth))
  for (const nodeType of query.node_types ?? []) parameters.append('node_types', nodeType)
  const suffix = parameters.size > 0 ? `?${parameters.toString()}` : ''
  return readJson(await fetch(`/api/research-lineage${suffix}`), 'GET /api/research-lineage') as Promise<ResearchLineageGraph>
}

export async function getResearchLineageSummary(): Promise<ResearchLineageSummary> {
  return readJson(await fetch('/api/research-lineage/summary'), 'GET /api/research-lineage/summary') as Promise<ResearchLineageSummary>
}

File: backend/app/api/research_lineage.py
from typing import Annotated

from fastapi import APIRouter, HTTPException, Query

from app.datasets import dataset_registry
from app.discovery import hypothesis_repository
from app.factor_relationships import factor_relationship_repository
from app.factors import factor_research_repository
from app.portfolio_lab.repository import portfolio_research_repository
from app.research_integrity import ResearchIntegrityEngine
from app.research_ledger import research_ledger
from app.research_lineage import (
    LineageDirection,
    LineageNodeType,
    ResearchLineageBuilder,
    ResearchLineageGraph,
    ResearchLineageService,
    ResearchLineageSummary,
)
from app.research_snapshots import research_snapshot_repository
from app.runs import run_store
from app.sdk.registry import strategy_registry
from app.strategy_drift import strategy_drift_repository
from app.walk_forward import walk_forward_repository
from app.workspaces import WorkspaceNotFoundError

from .workspaces import workspace_service

router = APIRouter(prefix="/api/research-lineage", tags=["research-lineage"])


def _service() -> ResearchLineageService:
    integrity = ResearchIntegrityEngine(
        dataset_registry,
        factor_research_repository,
        factor_relationship_repository,
        walk_forward_repository,
        hypothesis_repository,
        portfolio_research_repository,
        strategy_registry,
        run_store.repository,
        research_ledger,
    )
    return ResearchLineageService(
        ResearchLineageBuilder(
            dataset_registry,
            factor_research_repository,
            factor_relationship_repository,
            walk_forward_repository,
            portfolio_research_repository,
            hypothesis_repository,
            strategy_registry,
            run_store.repository,
            research_snapshot_repository,
            integrity,
            drift_reports=strategy_drift_repository,
        )
    )


@router.get("/summary", response_model=ResearchLineageSummary)
def research_lineage_summary() -> ResearchLineageSummary:
    return _service().summary()


@router.get("", response_model=ResearchLineageGraph)
def research_lineage(
    root_type: LineageNodeType | None = None,
    root_id: str | None = None,
    direction: LineageDirection = "BOTH",
    max_depth: Annotated[int, Query(ge=1, le=8)] = 8,
    node_types: Annotated[list[LineageNodeType] | None, Query()] = None,
    workspace_id: str | None = None,
) -> ResearchLineageGraph:
    try:
        members = (
            None
            if workspace_id is None
            else frozenset(
                (item.object_type, item.object_id)
                for item in workspace_service().memberships(workspace_id)
            )
        )
        return _service().graph(
            root_type=root_type,
            root_id=root_id,
            direction=direction,
            max_depth=max_depth,
            node_types=tuple(node_types or ()),
            workspace_members=members,
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except WorkspaceNotFoundError as exc:
        raise HTTPException(
            status_code=404, detail=f"Workspace '{exc.args[0]}' was not found"
        ) from exc
    except KeyError as exc:
        detail = f"Lineage root '{exc.args[0]}' was not found"
        raise HTTPException(status_code=404, detail=detail) from exc
